optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java

```diff
@@ -526,36 +526,382 @@ private static Runnable innerCountDistinctLong(CountDistinctLongResultContainer
+     *
+     * @param <A> type of the matched fact
+     * @param comparator never null
+     * @return never null
+     */
     public static <A> UniConstraintCollector<A, ?, A> min(Comparator<A> comparator) {
-        return minOrMax(comparator, true);
+        return min(Function.identity(), comparator);
     }
 
-    public static <A extends Comparable<A>> UniConstraintCollector<A, ?, A> min() {
-        return min(Comparable::compareTo);
+    /**
+     * Returns a collector that finds a minimum value in a group of elements, using the provided {@link Comparator}.
+     * <p>
+     *     It is expected that the {@link Comparator} is consistent with {@link #equals(Object)}.
+     *     In other words, if two elements compare to equal, any of them may be returned by the collector as they are
+     *     considered to be {@link #equals(Object)}.
+     * <p>
+     *     Example: Assume the following elements of type Person: Ann(age = 20), Beth(age = 25), Cathy(age = 30),
+     *     David (age = 25), Eric (age = 20).
+     *     Further assume that the {@link Comparator} provided imposes the usual increasing ordering on numbers.
+     *     Using Person::getAge as the groupValueMapping, the resulting collector returns 20, as that is the minimum
+     *     age of all the elements.
+     *
+     * @param <A> type of the matched fact
+     * @param <Mapped> type of the result
+     * @param groupValueMapping never null, maps facts from the matched type to the result type
+     * @param comparator never null
+     * @return never null
+     */
+    public static <A, Mapped> UniConstraintCollector<A, ?, Mapped> min(Function<A, Mapped> groupValueMapping,
+            Comparator<Mapped> comparator) {
+        return minOrMax(groupValueMapping, comparator, true);
+    }
+
+    /**
+     * Returns a collector that finds a minimum value in a group of {@link Comparable} elements.
+     * <p>
+     *     It is expected that the {@link Comparable}'s {@link Comparable#compareTo(Object)} is consistent with
+     *     {@link #equals(Object)}.
+     *     In other words, if two elements compare to equal, any of them may be returned by the collector as they are
+     *     considered to be {@link #equals(Object)}.
+     * <p>
+     *     Example: Assume the following mapped elements of type Person: Ann(age = 20), Beth(age = 25), Cathy(age = 30),
+     *     David (age = 25), Eric (age = 20).
+     *     Further assume that Person compares by age.
+     *     The resulting collector returns 20, as that is the minimum age of all the elements.
+     *
+     * @param <A> type of the first matched fact
+     * @param <B> type of the second matched fact
+     * @param <Mapped> type of the result
+     * @param groupValueMapping never null, maps matched facts to the result type
+     * @return never null
+     */
+    public static <A, B, Mapped extends Comparable<Mapped>> BiConstraintCollector<A, B, ?, Mapped> min(
```

Comment:
So, fun problem...

`ConstraintCollectors.min(BiFunction)`
`ConstraintCollectors.min(Comparator)`

These are ambiguous, for example:
`ConstraintCollectors.min((Integer a, Integer b) -> a + b);`
IDE doesn't know which one to pick.

What now?

- We can't remove the Comparator one, as that's old API.
- People could just cast to whichever they prefer, but that's ugly.

This is only going to be a problem with bi-stream. tri-stream and quad-stream are fine.
It's also only going to be a problem for `BiFunction<A, A, Integer>`, as that is ambiguous with `Comparator<A>`.

Considering all of that, I don't see it as a big enough problem to fix it.
Essentially the only solution is to rename `min(BiFunction)` to something like `minBi(...)`, and that's too draconian.

Reply:
-1 to minBi. (proposal C) The comparator one takes a back seat to the function one, not the other way around.
I am ok with casting for now. (proposal B)

Proposal A2: Deprecate the Comparator ones, add `minWithComparator()` and `maxWithComparator()`. Then we ditch the comparator for 8.0 (whenever that may be, but it could be sooner than later due to kogito), adding it in https://issues.redhat.com/browse/PLANNER-750 so we don't forget.